### PR TITLE
Add Graphql support to UserPickerField and support CustomUserSettings

### PR DESCRIPTION
**Pull request by hyzx86**
image

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Users/GraphQL/Startup.cs
Comment (hyzx86) [inline src/OrchardCore.Modules/OrchardCore.Users/GraphQL/Startup.cs:13]: Hi @gvkries ， Do you have any other suggestions on how to reuse this type, which would cause an error if you used Transient lifetime?

Could you help to Review this PR? Thanks!

**Comment by hyzx86** [comment]
@hishamco Could you help to review this PR? Thanks!

File: src/OrchardCore.Modules/OrchardCore.Users/GraphQL/Startup.cs
Comment (gvkries) [reply src/OrchardCore.Modules/OrchardCore.Users/GraphQL/Startup.cs:13]: The `UserType` is currently only populated by the `CurrentUserQuery` schema builder. It is not individually registered as a type in GraphQL, so I guess that has to change. I need to think about this...

**Comment by hishamco** [comment]
We will check once it's ready

**Comment by hyzx86** [comment]
Now I have a sequence-dependent problem. When I implement `UserPickerFieldGrpahType` in `ContentFields` Module , dynamic fields cannot be used because the `CustomUserSettings` in `UserType` has not been initialized.

However, if you migrate `UserPickerField` into the `User` module, I will also need to migrate all other resources for that field, such as some view files , `DispayDrivers`   and It also includes migration of some data structures, such as UserPickerFieldIndex

**Comment by hyzx86** [comment]
I updated the 'me' field with the `ContentItemInterface`

Everything seems normal so far

image

File: src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Startup.cs
Comment (gvkries) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Startup.cs:11]: I don't think this should depend on OrchardCore.Users, there is a dedicated `UserPickerStartup` class.

File: src/OrchardCore.Modules/OrchardCore.Users/GraphQL/Startup.cs
Comment (gvkries) [inline src/OrchardCore.Modules/OrchardCore.Users/GraphQL/Startup.cs:13]: Are you sure you get the right instance of the `UserType` in any circumstances? I think we should change the way the `UserType` is build, what I did within the `CurrentUserQuery` seems not enough anymore.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Startup.cs
Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Startup.cs:11]: Indeed, there is no need to depend on OrchardCore

File: src/OrchardCore.Modules/OrchardCore.Users/GraphQL/Startup.cs
Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.Users/GraphQL/Startup.cs:13]: Yes , Yes, they have the same hash code
image

**Comment by hyzx86** [comment]
Hi @gvkries  , do you have any other suggestions?

**Comment by hyzx86** [comment]
Hi @Skrypt ， Can you help to review this PR?

**Comment by hyzx86** [comment]
fixed: https://github.com/OrchardCMS/OrchardCore/issues/15540

**Comment by hyzx86** [comment]
@MikeAlhayek And this PR 🫡

**Comment by gvkries** [comment]
> Hi @gvkries , do you have any other suggestions?

Sorry, I didn't have time for this. I take a look as soon as possible.

**Comment by hyzx86** [comment]
Hi @hishamco , This PR is ready for review.

**Comment by gvkries** [comment]
@hyzx86 I created https://github.com/OrchardCMS/OrchardCore/pull/15691 to make the `UserType` more independent. With this, instead of injecting it to the `UserPickerFieldQueryObjectType`, you can then just let GraphQL resolve it:

E.g.
```
        public UserPickerFieldQueryObjectType()
        {
            Name = nameof(UserPickerField);

            Field<ListGraphType<StringGraphType>, IEnumerable<string>>("userIds")
                .Description("user ids")
                .PagingArguments()
                .Resolve(x =>
                {
                    return x.Page(x.Source.UserIds);
                });

            Field<ListGraphType<UserType>, IEnumerable<User>>("users")
                .Description("the user items")
                .PagingArguments()
                .ResolveAsync(x =>
                {
                    var userLoader = GetOrAddUserProfileByIdDataLoader(x);
                    return userLoader.LoadAsync(x.Page(x.Source.UserIds)).Then(itemResultSet =>
                    {
                        return itemResultSet.SelectMany(x => x);
                    });
                });

            Field<UserType, User>("user")
                .Description("the first user")
                .ResolveAsync(x =>
                {
                    var userLoader = GetOrAddUserProfileByIdDataLoader(x);
                    return userLoader.LoadAsync(x.Source.UserIds.FirstOrDefault()).Then(itemResultSet =>
                    {
                        return itemResultSet.FirstOrDefault();
                    });
                });
        }
```
This only requires the list to to be registered too:
```
services.AddTransient<ListGraphType<UserType>>();
```

That would avoid potential lifetime issues with a scoped DI registration.

**Comment by hyzx86** [comment]
depend on PR #15691

**Comment by Piedone** [comment]
@MikeAlhayek can you please elaborate on why do you think this needs triage? What should we be mindful of during the triage?

**Comment by MikeAlhayek** [comment]
It's one of those areas that I have not dig into much and no one else seems to want to review. So I added it so it can be reviewed and hopefully merged. Feel free to review it, remove the tag and merge it

**Comment by Piedone** [comment]
OK, thanks.

@gvkries do I understand correctly that your https://github.com/OrchardCMS/OrchardCore/pull/15691 addresses this issue as well, so we mustn't merge both?

**Comment by Piedone** [comment]
I see. What confused me is that that PR is supposed to fix https://github.com/OrchardCMS/OrchardCore/issues/15540 (as indicated in its description), just as this one (as indicated in the issue).

**Comment by gvkries** [comment]
> I see. What confused me is that that PR is supposed to fix #15540 (as indicated in its description), just as this one (as indicated in the issue).

Yes, that bug is fixed by both PRs, but this is more about the additional feature 🤗

**Comment by Piedone** [comment]
OK, thanks. I'll review this once https://github.com/OrchardCMS/OrchardCore/pull/15691 is merged, then.

**Comment by hishamco** [comment]
> OK, thanks. I'll review this once #15691 is merged, then.

It's merged

**Comment by Piedone** [comment]
Thanks! @hyzx86 please fix the merge conflict and let me know.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs
Comment (gvkries) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs:25]: I think all descriptions should be localizable.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs:27]: Use proper variable names instead of `x`, below too, and instead of `y` and `k` as well.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs:70]:
```suggestion
        }

        private static IDataLoader<string, IEnumerable<User>> GetOrAddUserProfileByIdDataLoader<T>(IResolveFieldContext<T> context)
        {
            var dataLoaderContextAccessor = context.RequestServices.GetRequiredService<IDataLoaderContextAccessor>();

            return dataLoaderContextAccessor.Context.GetOrAddCollectionBatchLoader("GetOrAddUserByIds", async (IEnumerable<string> userIds) =>
            {
                if (userIds == null || !userIds.Any())
                {
                    return null;
                }

                var session = context.RequestServices.GetService<ISession>();
                var users = await session.Query<User, UserIndex>(y => y.UserId.IsIn(userIds)).ListAsync();

                return users.ToLookup((User k) => k.UserId, (User user) => user);
            });
        }
```

File: src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs:55]: This will cause an `ArgumentNullException` if no user is picked in the field.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs:44]:
```suggestion
            Field<UserType, User>("firstUser")
```

File: src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs:53]: Do we really need this field?

**Review by Piedone — CHANGES_REQUESTED** [review]
When addressing review feedback, please adhere to the following:

- Please update your pull request according to feedback until it is approved by one of the core team members.
- Apply [suggested changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes) directly so the reviewer doesn't have to eyeball the changes. These resolve themselves after applying them, and that's fine.
- Don't resolve other conversations so it's easier to track for the reviewer. Then, the reviewer will resolve them.
- Feel free to mark conversations that you addressed to keep track of them with an emoji or otherwise, just don't resolve them.
- Please keep conversations happening in line comments in those convos, otherwise, communication will be a mess. If you have trouble finding them, see [this video](https://github.com/OrchardCMS/OrchardCore/pull/14749#issuecomment-1917976028).
- When you're done addressing all feedback of a review, click ["Re-request review"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#re-requesting-a-review) in the top-right corner for each reviewer when you're ready for another round of review, so they know that you're done.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs
Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs]: Why not? It can record the IDs of all content items and save them in a collection, and then query them in bulk from the database if the query need that field

Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs]: Oh sorry, I may have misunderstood you, you said 'first user', I think it should be useful in most cases, at least in js clients we can just use it without checking whether the array is empty first

``` js
const displayText=blog.author.firstUser?.displayText

const displayText2=blog.author?[0].displayText

```

File: src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs
Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs:64]: @Piedone , Can you give me a hint? thanks!

Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs:64]: > Handle potential null values gracefully to avoid `ArgumentNullException`.

@coderabbitai , Can you give me a hint?

File: src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs
Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs]: Done

File: src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs
Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs]: Done

File: src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs
Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs]: Done

File: src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs
Comment (gvkries) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs:41]:
```suggestion
                        return itemResultSet.SelectMany(users => users);
```

File: src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs
Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs]: Well, it's a bug left over from the previous version of the code, which I did in loader. Items that don't have this field return empty references, so I figured out a way to add a `firstUser` field 😅

image

Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Types/UserPickerFieldQueryObjectType.cs]: The firstUser field has been removed.

**Review by Piedone — APPROVED** [review]
Great, thank you!

**Pull request merged**
